### PR TITLE
Refactor rules

### LINKED ISSUE #367: Consider refactor of streamline rules.js, rules.html, rulescripts.html and subplanrules.html
Perhaps these files could be refactored or have diagram created to improve understanding.

### PR DESCRIPTION
**Pull request by j-selby**
This was painful...

Closes #367

This PR breaks apart rules into individual files, cleaning them up along the way. In doing this, I have attempted to remove duplicated logic, minimise the amount of information loaded for pages (i.e courses doesn't need program-specific rules), reduce LOCs across the board, and get rid of rulescripts entirely.

Further, I have tried to enforce ES6 throughout these new files.

This may break old course lists (due to change in name of vue object) - this was done to reduce ambiguity, but this needs to be verified independently.

### CODE REVIEW

**Comment by Daniel-Jang** [comment]
I was testing out the features and unfortunately, javascript spits out errors:
image

When you add a course rule in either or rule, the course rule is empty and the error appears (shown on the right side of the screenshot)

**Comment by Daniel-Jang** [comment]
Also, just a heads up, when you pull from develop (I have pulled both from develop and your branch to work on my "rule reordering" issue), you will see the following error when you add an 'either or' rule:

image

This is caused by the 'inconsistent_rules' counter/function to check either or rule group unit count inconsistencies.

**Comment by j-selby** [comment]
@Daniel-Jang the first one was something I missed - that should be easily fixable. The second one was because they didn't merge properly - I will rebase this after everyone gets their PRs through (else we will be here all night)

**Comment by Daniel-Jang** [comment]
> @Daniel-Jang the first one was something I missed - that should be easily fixable. The second one was because they didn't merge properly - I will rebase this after everyone gets their PRs through (else we will be here all night)

Ah okay, I'll keep following this branch as my branch currently depends on yours (you will have to merge yours in first).

**Comment by j-selby** [comment]
Rebased onto the latest develop branch.

File: cassdegrees/templates/staff/creation/createcourse.html [hunk 1]
```diff
@@ -84,7 +85,7 @@
 
     <script>
         function submit_form(form_action, redirect) {
-            if (handleRules()) {
+            if (app.export_rules()) {
```
Comment (JackSekoranja) [inline cassdegrees/templates/staff/creation/createcourse.html:88]:
```suggestion
            if (app.exportRules()) {
```

File: cassdegrees/templates/staff/creation/createprogram.html [hunk 2]
```diff
@@ -118,7 +119,7 @@
 
     <script>
         function submit_form(form_action, redirect) {
-            if (handleProgram() && handleRules()) {
+            if (globalRequirementsApp.export_requirements() && app.export_rules()) {
```
Comment (JackSekoranja) [inline cassdegrees/templates/staff/creation/createprogram.html:122]:
```suggestion
            if (globalRequirementsApp.export_requirements() && app.exportRules()) {
```

File: cassdegrees/templates/staff/creation/createsubplan.html [hunk 3]
```diff
@@ -91,12 +92,12 @@
         });
 
         function submit_course_form(){
-            handleRules();
+            app.export_rules();
```
Comment (JackSekoranja) [inline cassdegrees/templates/staff/creation/createsubplan.html:95]:
```suggestion
            app.exportRules();
```

File: cassdegrees/templates/staff/creation/createsubplan.html [hunk 4]
```diff
@@ -91,12 +92,12 @@
...
             document.getElementById('mainForm').submit();
         }
 
         function submit_form(redirect) {
-            if (handleProgram() && handleRules()) {
+            if (globalRequirementsApp.export_requirements() && app.export_rules()) {
```
Comment (JackSekoranja) [inline cassdegrees/templates/staff/creation/createsubplan.html:100]:
```suggestion
            if (globalRequirementsApp.export_requirements() && app.exportRules()) {
```

File: cassdegrees/templates/staff/creation/createprogram.html [hunk 5]
```diff
@@ -143,8 +144,8 @@
         }
 
         function force_submit_form(form_action, redirect) {
-            handleRules();
-            handleProgram();
+            app.export_rules();
```
Comment (JackSekoranja) [inline cassdegrees/templates/staff/creation/createprogram.html:147]:
```suggestion
            app.exportRules();
```

**Review by JackSekoranja — CHANGES_REQUESTED** [review]
After initial testing I found that I couldn't submit anything because the rules function is invalid (See suggestions). Note that if the following suggestions are added, there will be inconsistencies with the naming of export rules (i.e. `export_requirements` and `exportRules`), therefore we will need to decide whether we want to use camel case or snake case for our javascript functions.

**Comment by j-selby** [comment]
Didn't spot that one, sorry. Yeah - we will either need to change exportRules or all the other functions - this was mostly inconsistent before, though I was targeting underscores mostly.

**Comment by j-selby** [comment]
@JackSekoranja I've gone ahead and changed `exportRules` to be `export_rules` (for consistency with most of the other Vue rules) - might need to go through soon and fix the rest up (and maybe consider a Javascript linter?)

**Review by JackSekoranja — APPROVED** [review]
Can't find any issues. Happy for you to merge.

**Pull request merged**